core/trino-main/src/main/java/io/trino/sql/gen/columnar/DynamicPageFilter.java

```diff
@@ -0,0 +1,196 @@
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.trino.sql.gen.columnar.FilterEvaluator.createColumnarFilterEvaluator;
+import static io.trino.sql.ir.optimizer.IrExpressionOptimizer.newOptimizer;
+import static io.trino.sql.relational.SqlToRowExpressionTranslator.translate;
+import static java.util.Objects.requireNonNull;
+
+public final class DynamicPageFilter
+{
+    private final Metadata metadata;
+    private final TypeManager typeManager;
+    private final Session session;
+    private final IrExpressionOptimizer irExpressionOptimizer;
+    private final DomainTranslator domainTranslator;
+    private final DynamicFilter dynamicFilter;
+    private final Map<ColumnHandle, Symbol> columnHandles;
+    private final Map<Symbol, Integer> sourceLayout;
+    private final double selectivityThreshold;
+
+    @Nullable
+    @GuardedBy("this")
+    private Supplier<FilterEvaluator> compiledDynamicFilter;
+    @Nullable
+    @GuardedBy("this")
+    private CompletableFuture<?> isBlocked;
+
+    public DynamicPageFilter(
+            PlannerContext plannerContext,
+            Session session,
+            DynamicFilter dynamicFilter,
+            Map<Symbol, ColumnHandle> columnHandles,
+            Map<Symbol, Integer> sourceLayout,
+            double selectivityThreshold)
+    {
+        this.metadata = requireNonNull(plannerContext.getMetadata(), "metadata is null");
+        this.typeManager = requireNonNull(plannerContext.getTypeManager(), "typeManager is null");
+        this.session = requireNonNull(session, "session is null");
+        this.irExpressionOptimizer = newOptimizer(plannerContext);
+        this.domainTranslator = new DomainTranslator(plannerContext.getMetadata());
+        this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+        this.columnHandles = columnHandles.entrySet()
+                .stream()
+                .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+        this.sourceLayout = ImmutableMap.copyOf(sourceLayout);
+        this.selectivityThreshold = selectivityThreshold;
+        this.isBlocked = dynamicFilter.isBlocked();
+    }
+
+    // Compiled dynamic filter is fixed per-split and generated duration page source creation.
+    // Page source implementations may subsequently implement blocking on completion of dynamic filters, but since
+    // that occurs after page source creation, we cannot be guaranteed a completed dynamic filter here for initial splits
+    public synchronized Supplier<FilterEvaluator> createDynamicPageFilterEvaluator(ColumnarFilterCompiler compiler)
+    {
+        if (isBlocked == null) {
+            return compiledDynamicFilter;
+        }
+        // Recompile dynamic filter if it has changed since last compilation
+        if (compiledDynamicFilter == null || isBlocked.isDone()) {
+            isBlocked = dynamicFilter.isBlocked();
```

Comment:
nit: Ideally the order should be different
```
boolean isAwaitable = dynamicFilter.isAwaitable();
isBlocked = dynamicFilter.isBlocked();
```
since between `isBlocked()` and `isAwaitable()` there could be DF update and DF could stopped being awaitable.

This probably isn't a problem here since we check `getCurrentPredicate()` after `isAwaitable` call.

Reply:
In the current code "stopped being awaitable" means DF is complete (or there is no more lazy DFs and we anyway don't want to keep re-compiling for those).
In the below order
```
isBlocked = dynamicFilter.isBlocked();
boolean isAwaitable = dynamicFilter.isAwaitable();
```
If DF completes between `isBlocked()` and `isAwaitable()`, then we'll catch that and not re-compile again.
If DF updates between `isBlocked()` and `isAwaitable()` without completing, then we will just incur an unnecessary recompilation on next split, which is harmless.

In the opposite order
```
boolean isAwaitable = dynamicFilter.isAwaitable();
isBlocked = dynamicFilter.isBlocked();
```
If DF completes between `isAwaitable()` and `isBlocked()`, then will just incur an unnecessary recompilation on next split, which is harmless.
If DF updates between `isAwaitable()` and `isBlocked()` without completing, then will avoid the unnecessary recompilation on next split.

So both ways look equivalent to me, they'll incurr an extra re-compile in slightly different scenarios. I don't see strong reason to prefer one over the other.